Rename the key-pair elements returned in the gcryptGenerateX25519Keys method

javifernandez opened the pull request:
#### 33777aff18f133fb902b6f6b08438fe36f2bbeaa
<pre>
Rename the key-pair elements returned in the gcryptGenerateX25519Keys method
<a href="https://bugs.webkit.org/show_bug.cgi?id=289693">https://bugs.webkit.org/show_bug.cgi?id=289693</a>

Reviewed by Carlos Garcia Campos.

The gcryptGenerateX25519Keys method returns a std::pair of KeyMaterial variables
which should contain the public and private key. The caller expects the first
element to be the public key.

We were returning the data in the wrong order, so among other issues, it caused
that the public key generated by the JWK export of the private key doesn&apos;t match
the original key-pair&apos;s public key material.

This change fixes the issue by switching the variable names used to hold the
private and public key pair. This is also more consistent with the nomenclature
used in the Ed25519 generateKey() method..

Regarding tests, this change make the successes_X25519.https.any.html tests to
pass, so updating also the platform-specific TestExpectations file.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::gcryptGenerateX25519Keys):

Canonical link: <a href="https://commits.webkit.org/299198@main">https://commits.webkit.org/299198@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3a0e18f1fc0bc7dedc666eebc6d818547070334

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118226 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37904 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28543 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124385 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70267 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120104 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38597 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46486 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89713 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59350 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121179 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30751 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106008 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70209 "Found 1 new API test failure: TestWTF:WTF_Condition.OneProducerOneConsumerOneSlotTimeout (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29821 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24126 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68051 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100179 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24306 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127459 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45129 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34025 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98397 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45490 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102228 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98184 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24964 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43581 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21574 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41594 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45001 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50675 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44461 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47806 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46150 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -109,7 +109,7 @@ static std::optional<std::pair<Vector<uint8_t>, Vector<uint8_t>>> gcryptGenerate
         return std::nullopt;
 
-    return std::make_pair(WTFMove(*q), WTFMove(*d));
+    return std::make_pair(WTFMove(*d), WTFMove(*q));
```
there's probably a reason why these are called `d` and `q`, but it's not obvious. Would it make sense to rename them, or would that confuse people comfortable with crypto customs?

javifernandez replied:
The gcrypt nomenclature uses ```q``` for the public key and ```d``` for the private key.  It's also what it's used in the Ed25519 generateKey() method.

nikolaszimmermann replied:
I agree, we should use verbose names here. One shouldn't be a crypto expert to understand what is private and what is public key :-)

javifernandez replied:
As I said before, I don't disagree. However, I'd rather avoid doing "unrelated" changes all over the place to keep the consistency of the other cryptographic functions. I think that should be done in a different patch.

javifernandez committed:
Rename the key-pair elements returned in the gcryptGenerateX25519Keys method
https://bugs.webkit.org/show_bug.cgi?id=289693

Reviewed by Carlos Garcia Campos.

The gcryptGenerateX25519Keys method returns a std::pair of KeyMaterial variables
which should contain the public and private key. The caller expects the first
element to be the public key.

We were returning the data in the wrong order, so among other issues, it caused
that the public key generated by the JWK export of the private key doesn't match
the original key-pair's public key material.

This change fixes the issue by switching the variable names used to hold the
private and public key pair. This is also more consistent with the nomenclature
used in the Ed25519 generateKey() method..

Regarding tests, this change make the successes_X25519.https.any.html tests to
pass, so updating also the platform-specific TestExpectations file.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::gcryptGenerateX25519Keys):

Canonical link: https://commits.webkit.org/299198@main
